Provide an abstract base class to be able to test the formatter.

miklossy opened the pull request:
- Add the org.eclipse.xtext.testing.AbstractFormatterTest class.
- Modify the FormatterTest class of the Domainmodel example to demonstrate the usage of the AbstractFormatterTest class.
- Modify the AbstractXtendFormatterTest base class to inherit from the newly created AbstractFormatterTest class and adapt the Xtend formatter test cases to use the inherited members/methods.

Closes #2965

LorenzoBettini commented:
Just a comment about the CI failure: it looks like your PR is not based on the current master (it still uses Tycho 3 and old 4.31 builds.

miklossy commented:
Thanks for the hint @LorenzoBettini ! I have just rebased my feature branch on the master branch. Let's see if the CI bulid becomes green :-)

miklossy commented:
Seems the build is green now. @LorenzoBettini : could you please review my changes?

LorenzoBettini reviewed (approved):
I'm fine with this proposal, though I'd like to have confirmation also from @szarnekow 

Just a few other notes: I seem to remember that FormatterTestHelper provides other configurations. Do you think it could be useful to have additional assert methods in the base class accordingly?

Moreover, I think it would be important to use this base class also in other parts, e.g., in our formatter tests (Xbase, Xtend?) so that we're sure we use it during the CI. In fact, the Domainmodel is not tested in the CI of Xtext.

We could also have separate issues for the above.

miklossy commented:
Thanks for your feedback @LorenzoBettini !

> I'm fine with this proposal, though I'd like to have confirmation also from @szarnekow
I will request a review from @szarnekow as well.

> Just a few other notes: I seem to remember that FormatterTestHelper provides other configurations. Do you think it could be useful to have additional assert methods in the base class accordingly?
I updated the PR so that the base class contains methods to configure the formatter preferences.

> Moreover, I think it would be important to use this base class also in other parts, e.g., in our formatter tests (Xbase, Xtend?) so that we're sure we use it during the CI. In fact, the Domainmodel is not tested in the CI of Xtext.
I updated the Xtend Formatter test cases to use this new API class.

LorenzoBettini reviewed (approved):
I'm fine with the enhanced PR, thanks!

Just a comment/proposal: could we use the standard Java functional interfaces for the expected lambda instead of `Procedures.Procedure1`? I think `Consumer` would do, wouldn't it?

miklossy commented:
> Just a comment/proposal: could we use the standard Java functional interfaces for the expected lambda instead of Procedures.Procedure1? I think Consumer would do, wouldn't it?

Thanks for the feedback, @LorenzoBettini !
I have just updated the PR to use `java.util.function.Consumer` as suggested!

LorenzoBettini commented:
> I have just updated the PR to use `java.util.function.Consumer` as suggested!

@miklossy great! :)

miklossy commented:
Thanks for the review @LorenzoBettini : do you know where are the release notes prepared for Xtext 2.35 so that I can add some hints about this new API class?

LorenzoBettini commented:
> Thanks for the review @LorenzoBettini : do you know where are the release notes prepared for Xtext 2.35 so that I can add some hints about this new API class?

Thank you for the contribution @miklossy .

Concerning release notes, let's hear from @cdietrich

cdietrich commented:
please open a pr.

@LorenzoBettini can you create stubs for the new release notes

LorenzoBettini commented on org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.tests/xtend-gen/org/eclipse/xtext/example/domainmodel/tests/FormatterTest.java:
```diff
@@ -82,71 +72,62 @@ public void example1() {
+      it.<String>put(FormatterPreferenceKeys.indentation, " ");
     };
-    this._formatterTestHelper.assertFormatted(_function);
+    this.assertFormattedTo(_builder, _builder_1, _function);
```
@miklossy are you sure you committed all the generated files correctly? In my workspace, after using the new master with this PR merged, I get that this file should be

```java
    final Consumer<MapBasedPreferenceValues> _function = (MapBasedPreferenceValues it) -> {
```

instead of the committed

```java
    final Procedure1<MapBasedPreferenceValues> _function = (MapBasedPreferenceValues it) -> {
```

cdietrich replied:
i see these changes too. can you open a fixing pr

LorenzoBettini replied:
https://github.com/eclipse/xtext/pull/3003

LorenzoBettini commented:
> please open a pr.
> 
> @LorenzoBettini can you create stubs for the new release notes
> 
> see also #2948

@cdietrich done: https://github.com/eclipse/xtext-website/pull/33

hope I got it right because that's the first time

LorenzoBettini commented:
@miklossy the release notes files are in place https://github.com/eclipse/xtext-website/pull/33 if you want to add some hints about the new classes

miklossy commented:
Thank you @LorenzoBettini ! I will definitely do!

miklossy commented:
@LorenzoBettini : I created the PR https://github.com/eclipse/xtext-website/pull/34 about the hints.
